yellow-paper/docs/gas-and-fees/gas-metering-and-fees.md

```diff
@@ -0,0 +1,683 @@
+
+The public VM retains the current amount of `daGasLeft`, `l2GasLeft`, and `l1GasLeft` in its [MachineState](../public-vm/state.md#machine-state).
+
+This amount must be initialized for each public kernel iteration as:
+
+```
+daGasLeft := public_kernel_public_inputs.da_gas_limit 
+             - public_kernel_public_inputs.end_non_revertible.da_gas_used
+             - public_kernel_public_inputs.end.da_gas_used
+
+# analogous for l2GasLeft and l1GasLeft
+```
+
+At the end of execution, the public VM will have updated `daGasLeft`, `l2GasLeft`, and `l1GasLeft` values.
+
+This value will be used by the public kernel to update the `da_gas_used`, `l2_gas_used`, and `l1_gas_used` in the `PublicAccumulatedData` struct corresponding to the current phase:
+
+```
+# if in setup 
+public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used := 
+        public_kernel_public_inputs.gas_limits.da_gas_limit
+          - daGasLeft
+          - public_kernel_public_inputs.end.gas_used.da_gas_used
+
+# if in app logic
+public_kernel_public_inputs.end.gas_used.da_gas_used := 
+        public_kernel_public_inputs.gas_limits.da_gas_limit
+          - daGasLeft
+          - public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used
+
+# analogous for l2_gas_used and l1_gas_used
+```
+
+### Global gas limits
+
+The mechanics above imply that there will be a single global gas limit (for each gas dimension) for all enqueued public calls. However, users will still be able to specify individual gas limits for each *nested* call within an enqueued public call.
+
+Therefore, transactions will raise an OutOfGas **exception** if an operation would cause the `daGasLeft` or `l2GasLeft` to be negative, but this will only cause the transaction to **revert** if this occurs in the top-level call frame of an enqueued public call
+
+This global limit is shared across enqueued calls because unlike nested calls within the VM, there is no way to "catch" a revert of a top-level enqueued call, aside from running it as part of app logic.
+
+### Reverts in Setup/Teardown
+
+Setup and teardown are considered "non-revertible" phases. Thus, if a revert occurs in these phases (e.g the top level calls `REVERT` or divides by zero), the transaction is invalid and cannot be included in the block.
+
+For this reason, it is expected that sequencers will maintain a whitelist of setup and teardown functions, so that it can be quite sure that the transaction will not revert in these phases.
+
+### OutOfL2Gas reverts in App Logic
+
+If app logic reverts due to OutOfL2Gas, `l2GasLeft` is set to zero. Based on the update logic above, this results in 
+
+```
+public_kernel_public_inputs.gas_limits.l2_gas_limit ==
+  public_kernel_public_inputs.end.gas_used.l2_gas_used
+  + public_kernel_public_inputs.end_non_revertible.gas_used.l2_gas_used
+```
+
+Thus, the user is charged the full `l2_gas_limit`.
+
+However, we drop the DA and L1 gas used associated with app logic, i.e.:
```

Comment:
Is this because it would be unfair to charge these costs to the user, because these costs will never be borne by the sequencer/l1/da?

Reply:
Yes. I think it's more correct to say we drop all DA and L1 gas used values for **revertible** side-effects. This includes anything from private execution.